### PR TITLE
Create ExprProjectileBounceState.java

### PR DESCRIPTION
**Pull request by DeltaRays**
### Description

The title says it all
---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET || mode == ChangeMode.RESET) ? CollectionUtils.array(Boolean.class) : null;
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		boolean state = delta != null ? (Boolean) delta[0] : false;
+		for (Projectile entity : getExpr().getAll(e)) {
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:60]: Use getArray. Also, you can remove the braces on this for loop

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+	@Nullable
+	@Override
+	public Boolean convert(Projectile projectile) {
+		return projectile.doesBounce();
+	}
+	
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET || mode == ChangeMode.RESET) ? CollectionUtils.array(Boolean.class) : null;
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:54]: does this default to either true or false to all projectiles, if not, does Spigot/Paper provide a way to get each projectile's default bounce state? If not, reset changer makes little sense

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+@Name("Projectile Bounce State")
+@Description("A projectile's bounce state.")
+@Examples({"on projectile hit:",
+	"\tset projectile bounce state of event-projectile to true"})
+@Since("INSERT VERSION")
+public class ExprProjectileBounceState extends SimplePropertyExpression<Projectile, Boolean> {
+	
+	static {
+		register(ExprProjectileBounceState.class, Boolean.class, "[the] (projectile) bounce (state|ability|mode)", "projectiles");
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:42]: why is projectile on parenthesis?

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Projectile Bounce State")
+@Description("A projectile's bounce state.")
+@Examples({"on projectile hit:",
+	"\tset projectile bounce state of event-projectile to true"})
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:37]: I'd rather the example to be:
```suggestion
	"\tset bounce mode of shot projectile to true"})
```

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:37]: I don't want to remove 'projectile' as something regarding 'bounce' might be used in the future for something other than projectiles

**Comment by DeltaRays** [comment]
I've asked in the #skript-dev channel as to what I should use, effects & conditions or expressions

File: src/main/java/ch/njol/skript/conditions/CondProjectileCanBounce.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.SkriptParser.ParseResult;
+import ch.njol.util.Kleenean;
+
+@Name("Projectile Can Bounce")
+@Description("Whether or not a projectile can bounce.")
+@Examples({"on shoot:",
+	"\tsend \"Boing!\" to all players if projectile can bounce"})
+@Since("INSERT VERSION")
+public class CondProjectileCanBounce extends Condition {
```
Comment (FranKusmiruk) [inline src/main/java/ch/njol/skript/conditions/CondProjectileCanBounce.java:41]: This is a property condition, you may as well use the PropertyCondition class for it, with the CAN property type.

**Review by FranKusmiruk — DISMISSED** [review]

File: src/main/java/ch/njol/skript/conditions/CondProjectileCanBounce.java (hunk 5 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/conditions/CondProjectileCanBounce.java:41]: Okay, I'll use that

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+	public Class<? extends Boolean> getReturnType() {
+		return Boolean.class;
+	}
+	
+	@Override
+	protected String getPropertyName() {
+		return "projectile bounce state";
+	}
+	
+}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:74]: can you add a newline? :)

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:74]: Yeah, many of them would benefit from #3216

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+	
+	@Nullable
+	@Override
+	public Class<?>[] acceptChange(ChangeMode mode) {
+		return (mode == ChangeMode.SET) ? CollectionUtils.array(Boolean.class) : null;
+	}
+	
+	@Override
+	public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
+		boolean state = delta != null ? (Boolean) delta[0] : false;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:59]: I don't think you should do anything if delta is null

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java (hunk 7 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprProjectileBounceState.java:59]: 👍

**Review by APickledWalrus — DISMISSED** [review]
LGTM

**Review by Wealthyturtle — DISMISSED** [review]

**Review by ShaneBeee — APPROVED** [review]

**Pull request merged**